nebari-dev/nebari-self-registration
# Multi coupon support

## Description

This PR adds support for multiple coupons, so each can have independent configuration options. For example, having different coupons to add users to different keycloak groups.



## Related Issue
Closes #14 






## Motivation and Context

We've been using this extension in training and tutorials where there's a clear benefit of having different coupons map to different groups to manage access. For example, you might have one coupon granting access to GPUs, while other coupon just grants access to regular instances. Or you might have yet another coupon which grants admin rights.

Having support for multiple coupons make it easy to handle these cases by assigning them to different keycloak groups and managing separate permissions for them.




## How Has This Been Tested?

This has been tested by:
1. Installing the extension from the `multi-coupon` branch:
```bash
pip install git+https://github.com/marcelovilla/nebari-self-registration@multi-coupon
```

2. Set the following extension configuration in `nebari-config.yaml` and create the required groups in keycloak if they do not exist already:
```yaml
self_registration:
  coupons:
    METROSTAR:
      account_expiration_days: 365
      approved_domains: ["*.mil"]
      registration_groups: ["metrostar"]
    QUANSIGHT:
      account_expiration_days: 31
      approved_domains: ["quansight.com"]
      registration_groups: ["quansight"]
    METROSTARADMIN:
      account_expiration_days: 365
      approved_domains: ["*.mil"]
      registration_groups: ["metrostar", "admin"]
  values:
    image:
      repository: ghcr.io/marcelovilla/nebari-self-registration
      tag: "20240701-1553"
```

3. Re-deploy Nebari:
```bash
nebari deploy -c nebari-config.yaml
```

4. Go to `/registration` and use the different coupons to create new accounts. If everything works as expected, you'll see the new users in keycloak and depending what coupon you used, you'll see they've been assigned to the corresponding groups.






## Screenshots (if appropriate):
